conference/exchangerates.py

```diff
@@ -0,0 +1,125 @@
+
+Starting in 2018, with EuroPython in Edinburgh, we need to include VAT in local
+currency (GBP), to do that we can use ECB published exchange rates available at
+this URL
+    -> https://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml
+
+The way it works is this xml shows latest (published every day at 1600 CET) ECB
+exchange rates.
+
+What we're going to do is take that XML, parse it, and then cache the results
+for 24 hours, because we're just interested in using latest value (if we need
+historical values we can get full XML since 1999 here:
+
+    https://www.ecb.europa.eu/stats/eurofxref/eurofxref-hist.xml
+    or as CSV
+    https://www.ecb.europa.eu/stats/eurofxref/eurofxref-hist.zip
+
+------------------------------------------------------------------------------
+THIS WHOLE MODULE ASSUMES WE'RE BASING EVERYTHING IN EUROS,
+AND CONVERT EITHER TO OR FROM EUROS. KEEP THAT IN MIND.
+------------------------------------------------------------------------------
+"""
+
+from __future__ import unicode_literals, absolute_import, print_function
+
+from datetime import datetime, date
+from decimal import Decimal
+
+from django.core.cache import cache
+
+import requests
+from lxml import etree as ET
+
+
+DAILY_ECB_URL = "https://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml"
+CURRENCY_CACHE_KEY = 'currency_xrates'
+CURRENCY_CACHE_TIMEOUT = 60 * 60 * 24  # 24 hours
+
+# we probably can assume rounding to 0.01 for most of the currencie we're going
+# to ever use.
+DEFAULT_DECIMAL_PLACES = Decimal('0.01')
+
+# The actual XML will have more currencies.
+EXAMPLE_ECB_DATE = date(2018, 3, 6)
+EXAMPLE_ECB_DAILY_XML = """
+<?xml version="1.0" encoding="UTF-8"?>
+<gesmes:Envelope xmlns:gesmes="http://www.gesmes.org/xml/2002-08-01"
+    xmlns="http://www.ecb.int/vocabulary/2002-08-01/eurofxref">
+    <gesmes:subject>Reference rates</gesmes:subject>
+    <gesmes:Sender>
+        <gesmes:name>European Central Bank</gesmes:name>
+    </gesmes:Sender>
+    <Cube>
+        <Cube time="2018-03-06">
+            <Cube currency="USD" rate="1.2411"/>
+            <Cube currency="JPY" rate="131.84"/>
+            <Cube currency="GBP" rate="0.89165"/>
+        </Cube>
+    </Cube>
+</gesmes:Envelope>
+""".strip()
+
+
+def normalize_price(price):
+    """
+    TODO: move this function somewhere else, it's not a most obvious place to
+    put it.
+    """
+    return price.quantize(Decimal(DEFAULT_DECIMAL_PLACES))
+
+
+def fetch_latest_ecb_exrates():
+    """
+    Example of the XML (see EXAMPLE_XML constant)
+    """
+    response = requests.get(DAILY_ECB_URL)
+    info = ET.fromstring(response.content)[2][0]
+    datestamp = datetime.strptime(info.attrib['time'], "%Y-%m-%d").date()
+    rates = [x.attrib for x in info]
+
+    return dict(
+        datestamp=datestamp,
+        **{x['currency']: Decimal(x['rate']) for x in rates}
+    )
+
+
+def get_ecb_rates_for_currency(currency):
```

Comment:
@malemburg @matrixise (re the comment above about 404) – so this is the function that you supposed to use to actually get the rate – and `fetch_` is basically just low-level function to just the fetching – however we should probably add at least some assert on the response_code for fetch, so in case of failure (for example because ECB rate provider is down, or broken) we at least get a meaningful error message (and not some "xml is invalid" or something like that). What do you think?

Reply:
See above: asserts should not be used for detecting production error situations. They are only meant for development level check which don't occur in production. Please use proper error handling and use a fall-back solution (e.g. reuse the last cached value) to work around network issues or failures on the ECB side.